### ISSUE TITLE
ForbiddenVoid and TestObserver<Void>

### ISSUE THREAD

**Issue by realdadfish** [arturbosch/detekt#1724]
This is a follow-up on #1472 - I just stumbled upon the very same issue with RxJava and TestObserver of a Completable:

```kotlin
fun testMe(): TestObserver<Void> = Completable.complete().test()
```

is then also flagged with ForbiddenVoid. IMHO this rule should get some configuration, such as onlyCheckTypesInPackage which we would naturally configure to our project package path and the rule could then ignore types that reside in different (base) packages from that.

**Comment by schalkms** [comment]
I think this is a classical edge case for suppressing this one concern.
I would favor to not include any workarounds for this.
Why don't you suppress this?
Another idea would be to write a custom rule, which basically copies detekt's default one and includes your check and config option.

**Comment by realdadfish** [comment]
Suppressing this one concern means suppressing it in many different test cases where TestObserver is used on a Completable. I'd rather disable the rule altogether than suppressing single occurrences. And I don't know how common it is for people to copy existing rules into their own implementations, I'd favour a configurable upstream implementation from which everybody could benefit.

**Comment by schalkms** [comment]
> Suppressing this one concern means suppressing it in many different test cases where TestObserver is used on a Completable. I'd rather disable the rule altogether than suppressing single occurrences.

You use this quite frequently. Hence, I suggest to deactivate the rule.

> I don't know how common it is for people to copy existing rules into their own implementations

Yeah, that's no problem. We can't cover every (domain-specific) use-case in every rule. This would lead to a higher complexity in detekt.

The only solution I see here is to add a config option named `ignoreUsageInGenerics` or something similar. Would this config option help you in your case?

**Comment by realdadfish** [comment]
Yes, this could help if this is also easy to implement. I mean, in the end the issue is an interop one. I'd personally not use `Void` anywhere myself, unless I'm forced to do so because I'm doing interop with a Java-only library. That's why I originally had the `onlyCheckTypesInPackage` configuration idea. RxJava's `TestObserver` naturally does not live in my application's package, so it could be ignored - just like all else.

**Comment by schalkms** [comment]
Thanks for the explanation.
Do you use `Void` only a generic type parameter?

**Comment by realdadfish** [comment]
Yes, I don't use `Void` anywhere else.

**Comment by schalkms** [comment]
Then I would go with the `ignoreUsageInGenerics` config option.
There should be a condition which checks for the occurrence of `Void` inside a generic type declaration (inside the angle brackets <>).
https://github.com/arturbosch/detekt/blob/13fc590ff306ed5d09ec15f27e81587d813f903e/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenVoid.kt#L50

**Comment by schalkms** [comment]
Could you @realdadfish help us out here?

**Comment by realdadfish** [comment]
I'd love to, but is there any documentation available as to how to properly set up one's working environment? I have build issues in both, AS and IJ. The Spek plugin that I need to run tests only seems to work in my AS 3.4 installation. What is the suggested developement environment for Detekt?

**Comment by schalkms** [comment]
Nice!
I use the [IntelliJ Spek plugin](https://plugins.jetbrains.com/plugin/8564-spek), which I installed from the official plugin repository.
What errors do you get in IntelliJ?

Our [contribution section](https://github.com/arturbosch/detekt/blob/master/.github/CONTRIBUTING.md) describes what needs to be done in order to setup detekt.

**Comment by realdadfish** [comment]
I tried this plugin, but I'm already on AS 3.4 and cannot install this plugin there. I think I'll update my IJ 2018.2 to 2018.3 and try this artefact then. I also tried https://plugins.jetbrains.com/plugin/10915-spek-framework and this seems to install fine in AS 3.4, but I had issues there (`org.intellij` imports could not be resolved and amongst other things). I'll check the contribution section for more hints.

**Comment by schalkms** [comment]
Interesting..
Regarding the `org.intellij` - did you try do execute the gradle clean task?
This seems very strange.
If you need further help, please feel free to ask us.

**Comment by realdadfish** [comment]
Got it working with IJ 2019.1.3 and the Spek plugin.
